### ISSUE TITLE
Sentiment Analysis

### ISSUE THREAD

**Issue by cddesire** [cddesire/cddesire.github.io#31]
### 3、Summary the sentiment analysis on deep learning

**Comment by cddesire** [comment]
## tf.clip_by_global_norm 
修正梯度值，用于控制梯度爆炸的问题，需要对梯度进行修剪。

``` python
tf.clip_by_global_norm(t_list, clip_norm, use_norm=None, name=None) 
```
`t_list`是梯度张量， `clip_norm`是截取的比率clipping ratio。
``` python
t_list[i] * clip_norm / max(global_norm, clip_norm)
global_norm = sqrt(sum([l2norm(t)**2 for t in t_list]))
```
当t_list的L2模大于指定的global_norm时，就会对t_list做等比例缩放

**Comment by cddesire** [comment]
## tf.stack tf.unstack
``` python
import tensorflow as tf
a = tf.constant([1,2,3])
b = tf.constant([4,5,6])
c = tf.stack([a,b],axis=1)
d = tf.unstack(c,axis=0)
e = tf.unstack(c,axis=1)
print(c.get_shape())
with tf.Session() as sess:
    print(sess.run(c))
    print(sess.run(d))
    print(sess.run(e))
```
(3, 2)

[[1 4] 
[2 5] 
[3 6]]

[array([1, 4]), array([2, 5]), array([3, 6])]

[array([1, 2, 3]), array([4, 5, 6])]

**Comment by cddesire** [comment]
## tf.control_dependencies
``` python
with g.control_dependencies([a, b, c]):
  # `d` and `e` will only run after `a`, `b`, and `c` have executed.
  d = ...
  e = ...
```

**Comment by cddesire** [comment]
``` python
tf.train.range_input_producer(limit, num_epochs=None, 
shuffle=True, seed=None, capacity=32, 
shared_name=None, name=None)
```
产生一个从1至limit-1的整数至队列中

**Comment by cddesire** [comment]
## tf.contrib.rnn.BasicLSTMCell
``` python
tf.contrib.rnn.BasicLSTMCell(size, forget_bias=0.0, state_is_tuple=True)
```

**Comment by cddesire** [comment]
## tf.gradients
``` python
def gradients(ys, xs)
```
ys和xs都可以是一个tensor或者tensor列表。而计算完成以后，该函数会返回一个长为len(xs)的tensor列表，列表中的每个tensor是ys中每个值对xs[i]求导之和

**Comment by cddesire** [comment]
## tf.dynamic_rnn
``` python
tf.nn.dynamic_rnn(cell, inputs)
```
假设RNN的输入input是[2,20,128]，其中2是batch_size，20是文本最大长度，128是embedding_size。dynamic_rnn返回的是两个参数：outputs，last_states，其中outputs是[2,20,128]，也就是每一个迭代隐状态的输出，last_states是由(c，h)组成的tuple，均为[batch,128]。

**Comment by cddesire** [comment]
https://zhuanlan.zhihu.com/p/26996066
https://zhuanlan.zhihu.com/p/27552230
https://zhuanlan.zhihu.com/p/27485097
